client/service/src/lib.rs

```diff
@@ -332,6 +373,44 @@ async fn build_network_future<
 	}
 }
 
+fn build_hardcoded_sync<TBl, TCl, TBackend>(
+	client: Arc<TCl>,
+	backend: Arc<TBackend>,
+) -> Result<sc_chain_spec::HardcodedSync<TBl>, sc_rpc::system::error::Error>
+	where
+		TBl: BlockT,
+		TCl: HeaderBackend<TBl>,
+		TBackend: MaybeChtRootStorageProvider<TBl>,
+{
+	let storage = backend.cht_root_storage().ok_or(sc_rpc::system::error::Error::BackendNoChtRoots)?;
+
+	let finalized_hash = client.info().finalized_hash;
+	let finalized_number = client.info().finalized_number;
+
+	use sc_client_api::cht;
+
+	let mut chts = Vec::new();
+
+	/// We can't fetch a CHT root later than `finalized_number - 2 * cht_size`.
+	let cht_size_x_2 = cht::size::<NumberFor::<TBl>>() * NumberFor::<TBl>::from(2);
+
+	let mut number = NumberFor::<TBl>::one();
+
+	while number <= finalized_number.saturating_sub(cht_size_x_2) {
+		match storage.header_cht_root(cht::size(), number)? {
+			Some(cht_root) => chts.push(cht_root),
+			None => log::error!("No CHT found for block {}", number),
```

Comment:
In theory, `storage.header_cht_root` should never return `None` for this range, but in practise I think it might be better to keep this here instead of panicking.